### ISSUE TITLE
ci-instability: connect and stream cortex ingester pod logs from /loki/api/v1/tail: Uncaught Error: Unexpected server response: 500

### ISSUE THREAD

**Issue by sreis** [opstrace/opstrace#1334]
**Describe the bug**

```
[2021-08-26T21:32:40Z]   1) Loki tail API test suite
[2021-08-26T21:32:40Z]        connect and stream cortex ingester pod logs from /loki/api/v1/tail:
[2021-08-26T21:32:40Z]      Uncaught Error: Unexpected server response: 500
[2021-08-26T21:32:40Z]       at ClientRequest.<anonymous> (node_modules/ws/lib/websocket.js:750:7)
[2021-08-26T21:32:40Z]       at ClientRequest.emit (node:events:394:28)
[2021-08-26T21:32:40Z]       at ClientRequest.emit (node:domain:475:12)
[2021-08-26T21:32:40Z]       at HTTPParser.parserOnIncomingClient [as onIncoming] (node:_http_client:621:27)
[2021-08-26T21:32:40Z]       at HTTPParser.parserOnHeadersComplete (node:_http_common:128:17)
[2021-08-26T21:32:40Z]       at HTTPParser.execute (<anonymous>)
[2021-08-26T21:32:40Z]       at TLSSocket.socketOnData (node:_http_client:487:22)
[2021-08-26T21:32:40Z]       at TLSSocket.emit (node:events:394:28)
[2021-08-26T21:32:40Z]       at TLSSocket.emit (node:domain:475:12)
[2021-08-26T21:32:40Z]       at addChunk (node:internal/streams/readable:315:12)
[2021-08-26T21:32:40Z]       at readableAddChunk (node:internal/streams/readable:289:9)
[2021-08-26T21:32:40Z]       at TLSSocket.Readable.push (node:internal/streams/readable:228:10)
[2021-08-26T21:32:40Z]       at TLSWrap.onStreamRead (node:internal/stream_base_commons:199:23)
[2021-08-26T21:32:40Z] 
[2021-08-26T21:32:40Z]   2) Loki tail API test suite
[2021-08-26T21:32:40Z]        connect and stream cortex ingester pod logs from /loki/api/v1/tail:
[2021-08-26T21:32:40Z]      done() called multiple times in test <Loki tail API test suite connect and stream cortex ingester pod logs from /loki/api/v1/tail> of file /build/test/test-remote/test_loki_tail_endpoint.ts; in addition, done() received error: Error: Promise rejected with no or falsy reason
[2021-08-26T21:32:40Z]     at /build/node_modules/mocha/lib/runnable.js:377:26
[2021-08-26T21:32:40Z]     at runMicrotasks (<anonymous>)
[2021-08-26T21:32:40Z]     at processTicksAndRejections (node:internal/process/task_queues:96:5)
[2021-08-26T21:32:40Z]   Error: done() called multiple times in test <Loki tail API test suite connect and stream cortex ingester pod logs from /loki/api/v1/tail> of file /build/test/test-remote/test_loki_tail_endpoint.ts; in addition, done() received error: Error: Promise rejected with no or falsy reason
[2021-08-26T21:32:40Z]       at runMicrotasks (<anonymous>)
[2021-08-26T21:32:40Z]       at processTicksAndRejections (node:internal/process/task_queues:96:5)
[2021-08-26T21:32:40Z]       at runMicrotasks (<anonymous>)
[2021-08-26T21:32:40Z]       at processTicksAndRejections (node:internal/process/task_queues:96:5)
```

https://buildkite.com/opstrace/upgrade-tests/builds/1270#36c87cfe-7c90-4683-9e42-7bdd56e4b125
https://buildkite.com/opstrace/upgrade-tests/builds/1271#82db6e96-a6c5-464f-ae53-684e0b4a3e03

**Comment by nickbp** [comment]
https://buildkite.com/opstrace/upgrade-tests/builds/1275#f29d2d73-1d81-42c7-8f26-9acc9f4ffebf/1317-3335
https://buildkite.com/opstrace/upgrade-tests/builds/1276#a38df4f6-337d-4b8c-af56-2a3fa0ff6bd8/6-1422
https://buildkite.com/opstrace/upgrade-tests/builds/1278#3e3a007b-c223-4bf6-a280-40c42d971838/6-1690
https://buildkite.com/opstrace/upgrade-tests/builds/1279#34d3a29e-feca-475d-9a4f-87b4dab1baec/228-2238
[... skipped some for brevity ...]
https://buildkite.com/opstrace/upgrade-tests/builds/1291#67a71e82-520f-483b-a20b-a7b850ce5aeb/6-1823

**Comment by nickbp** [comment]
main build's embedded upgrade test: https://buildkite.com/opstrace/scheduled-main-builds/builds/2977#5f1b0af6-19e1-4d3d-8a2c-912cfc8f841d/4335-6345

**Comment by nickbp** [comment]
main build's embedded upgrade test: https://buildkite.com/opstrace/scheduled-main-builds/builds/2982#c1bbd229-cde3-4796-83ab-95e9cc185fd2/4145-6146

**Comment by nickbp** [comment]
main build's embedded upgrade test (also failed regular AWS but that's a different failure): https://buildkite.com/opstrace/scheduled-main-builds/builds/2983#0a79200f-3ddd-4abc-93c8-d41f6cebab97/4028-6040

**Comment by nickbp** [comment]
embedded upgrade test: https://buildkite.com/opstrace/scheduled-main-builds/builds/2985#6939fe35-915d-49ed-81bd-82ee7c85ff4e/4179-6189

**Comment by nickbp** [comment]
embedded upgrade test: https://buildkite.com/opstrace/scheduled-main-builds/builds/2986#9ea02188-eecd-403a-9111-41179c17d578/4101-6109

**Comment by nickbp** [comment]
upgrade tests continuing: https://buildkite.com/opstrace/upgrade-tests/builds/1292#8283ba20-1784-4c18-a5f1-33901c507895/1871-3886

**Comment by sreis** [comment]
https://buildkite.com/opstrace/scheduled-main-builds/builds/2987#7c81515a-29e2-4c44-9896-8fd4a6667976

**Comment by sreis** [comment]
https://buildkite.com/opstrace/scheduled-main-builds/builds/2988#2c170ad8-5799-4343-9162-100b48ac333f/863-2873

This time it timed out in the regular AWS test: 

```
[2021-08-30T06:50:27Z]   1) Loki tail API test suite
[2021-08-30T06:50:27Z]        connect and stream cortex ingester pod logs from /loki/api/v1/tail:
[2021-08-30T06:50:27Z]
[2021-08-30T06:50:27Z]       AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:
[2021-08-30T06:50:27Z] + actual - expected
[2021-08-30T06:50:27Z]
[2021-08-30T06:50:27Z] + 'timeout'
[2021-08-30T06:50:27Z] - 'test'
[2021-08-30T06:50:27Z]       + expected - actual
[2021-08-30T06:50:27Z]
[2021-08-30T06:50:27Z]       -timeout
[2021-08-30T06:50:27Z]       +test
[2021-08-30T06:50:27Z]
[2021-08-30T06:50:27Z]       at Context.<anonymous> (test_loki_tail_endpoint.ts:79:12)
```

**Comment by sreis** [comment]
https://buildkite.com/opstrace/upgrade-tests/builds/1293#52d4e511-7ad1-455d-8b20-c45b2e1d0a07

**Comment by sreis** [comment]
https://buildkite.com/opstrace/upgrade-tests/builds/1294

**Comment by sreis** [comment]
Ran an upgrade manually from version v2021.08.13 to the latest main. After the upgrade, I could connect via `wscat`:

```
$ wscat -n -H "Authorization: Bearer $(cat tenant-api-token-system)" --connect "wss://loki.system.sreis.opstrace.io/loki/api/v1/tail?query=%7Bk8s_namespace_name%3D%22cortex%22%2Ck8s_container_name%3D%22ingester%22%7D"
Connected (press CTRL+C to quit)                                                                                                                       
< {"streams":[{"stream":{"k8s_container_name":"ingester","k8s_host":"ip-192-168-49-62.us-west-2.compute.internal","k8s_namespace_name":"cortex","k8s_pod_name":"ingester-1"},"values":[["1630315091404934150","log=\"level=info ts=2021-08-30T09:18:11.404746169Z caller=main.go:193 msg=\\\"Starting Cortex\\\" version=\\\"(version=1.10.0, branch=master, revision=e658571)\\\"\n\" stream=stderr time=2021-08-30T09:18:11.40493415Z"]]},{"stream":{"k8s_container_name":"ingester","k8s_host":"ip-192-168-49-62.us-west-2.compute.internal","k8s_namespace_name":"cortex","k8s_pod_name":"ingester-1"},"values":[["163031509
```

**Comment by sreis** [comment]
Patched the tests locally to focus on the loki tail API endpoint tests:
```
> git diff
diff --git test/test-remote/test_loki_tail_endpoint.ts test/test-remote/test_loki_tail_endpoint.ts
index c8633882..15cef81f 100644
--- test/test-remote/test_loki_tail_endpoint.ts
+++ test/test-remote/test_loki_tail_endpoint.ts
@@ -28,7 +28,7 @@ import {
   enrichHeadersWithAuthTokenFile
 } from "./testutils";
 
-suite("Loki tail API test suite", function () {
+suite.only("Loki tail API test suite", function () {
   suiteSetup(async function () {
     log.info("suite setup");
     globalTestSuiteSetupOnce();

```

Ran test-remote and it passed:

```
2021-08-30T10:17:11.582Z info: got a message over websocket: {"streams":[{"stream":{"k8s_container_name":"ingester","k8s_host":"ip-192-168-49-62.us-west-2.compute.internal","k8s_namespace_name":"cortex","k8s_pod_name":"ingester-1"},"values":
    ✓ connect and stream cortex ingester pod logs from /loki/api/v1/tail (890ms)
2021-08-30T10:17:11.583Z info: suite teardown


  1 passing (898ms)

Done in 38.33s
```

**Comment by sreis** [comment]
After running the above tests I started to see the flaky behavior while trying to connect to the tail endpoint using wscat. Here's a sample:

```
> wscat -n -H "Authorization: Bearer $(cat tenant-api-token-system)" --connect "wss://loki.system.sreis.opstrace.io/loki/api/v1/tail?query=%7Bk8s_namespace_name%3D%22cortex%22%2Ck8s_container_name%3D%22ingester%22%7D"                                                                                      
Connected (press CTRL+C to quit)

> wscat -n -H "Authorization: Bearer $(cat tenant-api-token-system)" --connect "wss://loki.system.sreis.opstrace.io/loki/api/v1/tail?query=%7Bk8s_namespace_name%3D%22cortex%22%2Ck8s_container_name%3D%22ingester%22%7D"                                                                                      
error: Unexpected server response: 500                                                                                                                  > %                                                                                                                                                     

> wscat -n -H "Authorization: Bearer $(cat tenant-api-token-system)" --connect "wss://loki.system.sreis.opstrace.io/loki/api/v1/tail?query=%7Bk8s_namespace_name%3D%22cortex%22%2Ck8s_container_name%3D%22ingester%22%7D"                                                                                       
error: Unexpected server response: 500
```

The first request connected but the following two failed. 

Inspecting the querier logs further I found this error message printed in the querier logs:
 
```
"rpc error: code = Unknown desc = closing transport due to: connection error: desc = \"error reading from server: EOF\", received prior goaway: code: ENHANCE_YOUR_CALM, debug data: too_many_pings"
```

This seems like an issue with the [grpc-go connection between the querier and the query-frontend](https://github.com/grpc/grpc-go/issues/4266) for this there's not a lot of information or resolution yet.

To work around it I'll make the loki-api proxy the tail endpoint requests directly to the querier. This was how it was working before we introduced the query-frontend.

**Comment by nickbp** [comment]
Sounds like the issue's been found (:tada:) but recording these failures just for completeness' sake:
- https://buildkite.com/opstrace/upgrade-tests/builds/1298#ed5f17d5-408c-4bd1-8608-d0b7fdd5ad17/1074-3100
- https://buildkite.com/opstrace/scheduled-main-builds/builds/2992#e2d4efc4-3284-409c-bbc4-fc93585a0c39/4073-6085
- https://buildkite.com/opstrace/upgrade-tests/builds/1300#2691d308-9f32-4b8b-8a76-b0f3a5194f44/2487-4503

**Comment by sreis** [comment]
I'm working on a fix in PR #1340 but meanwhile, let's skip this particular test because the noise is getting out of hand.

**Comment by triclambert** [comment]
Buried amid a build with a GCP project failure was also this error:

```
[2021-08-31T09:37:45Z]   2 failing
[2021-08-31T09:37:45Z]
[2021-08-31T09:37:45Z]   1) Loki tail API test suite
[2021-08-31T09:37:45Z]        connect and stream cortex ingester pod logs from /loki/api/v1/tail:
[2021-08-31T09:37:45Z]      Uncaught Error: Unexpected server response: 500
[2021-08-31T09:37:45Z]       at ClientRequest.<anonymous> (node_modules/ws/lib/websocket.js:753:7)
[2021-08-31T09:37:45Z]       at ClientRequest.emit (node:events:394:28)
[2021-08-31T09:37:45Z]       at ClientRequest.emit (node:domain:475:12)
[2021-08-31T09:37:45Z]       at HTTPParser.parserOnIncomingClient [as onIncoming] (node:_http_client:621:27)
[2021-08-31T09:37:45Z]       at HTTPParser.parserOnHeadersComplete (node:_http_common:128:17)
[2021-08-31T09:37:45Z]       at HTTPParser.execute (<anonymous>)
[2021-08-31T09:37:45Z]       at TLSSocket.socketOnData (node:_http_client:487:22)
[2021-08-31T09:37:45Z]       at TLSSocket.emit (node:events:394:28)
[2021-08-31T09:37:45Z]       at TLSSocket.emit (node:domain:475:12)
[2021-08-31T09:37:45Z]       at addChunk (node:internal/streams/readable:315:12)
[2021-08-31T09:37:45Z]       at readableAddChunk (node:internal/streams/readable:289:9)
[2021-08-31T09:37:45Z]       at TLSSocket.Readable.push (node:internal/streams/readable:228:10)
[2021-08-31T09:37:45Z]       at TLSWrap.onStreamRead (node:internal/stream_base_commons:199:23)
[2021-08-31T09:37:45Z] 
[2021-08-31T09:37:45Z]   2) Loki tail API test suite
[2021-08-31T09:37:45Z]        connect and stream cortex ingester pod logs from /loki/api/v1/tail:
[2021-08-31T09:37:45Z]      Error: done() called multiple times in test <Loki tail API test suite connect and stream cortex ingester pod logs from /loki/api/v1/tail> of file /build/test/test-remote/test_loki_tail_endpoint.ts; in addition, done() received error: websocket closed
[2021-08-31T09:37:45Z]       at runMicrotasks (<anonymous>)
[2021-08-31T09:37:45Z]       at processTicksAndRejections (node:internal/process/task_queues:96:5)
[2021-08-31T09:37:45Z] 
```

https://buildkite.com/opstrace/upgrade-tests/builds/1303#988de0fd-1519-4927-b541-b9219e4a141d/6-1798
